Msvc2017 for merge

ericmehl opened the pull request:
Add Windows support to Cortex.

This is mostly a collection of small changes to fix specific issues with MSVC compatibility.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/ImageEngine/cortex/blob/master/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Cortex project's prevailing coding style and conventions.
- [X] If my code made breaking changes, I applied the **pr-majorVersion** label to this PR.

ericmehl commented:
The main outstanding item is getting all the tests running and passing which I'm working on separately and have planned for a future PR.

Please ask away with any questions, tips, requests and such.

johnhaddon commented on include/IECore/Canceller.h:
```diff
@@ -76,26 +76,11 @@ class IECORE_API Canceller : public boost::noncopyable
-		}
+		void cancel();
 
+		static void check(const Canceller *canceller);
```
I'm concerned about this no longer being inlined. We call `check()` from various performance-critical inner loops, and I did a fair bit of testing to ensure that it didn't have too much overhead. Not inlining potentially changes the balance there - is there some other way we can address the MSVC issue?

ericmehl replied:
Looks like an easy fix of just removing the export declaration which allows it to be inlined as intended.

johnhaddon commented on python/IECore/ClassLoader.py:
```diff
@@ -215,7 +215,7 @@ def defaultLoader( cls, envVar ) :
 				msg( Msg.Level.Warning, "ClassLoader.defaultLoader", "Environment variable %s not set." % envVar )
 
-			loader = cls( SearchPath( os.path.expandvars( sp ) ) )
+			loader = cls( SearchPath( os.path.expandvars( sp ), os.pathsep ) )
```
Is this definitely necessary? I think it should have been covered already by 49c6d73ff8dbe49748ebf673230ae1f680868b6f.

ericmehl replied:
Ah, wonderful, I hadn't seen that commit. Building and testing now without that explicit os.pathsep.

johnhaddon commented on python/IECore/ClassLsOp.py:
```diff
@@ -102,9 +102,9 @@ def doOperation( self, operands ) :
 
 			if operands["searchPath"].value :
-				sp = IECore.SearchPath( operands["searchPath"].value, ":" )
+				sp = IECore.SearchPath( operands["searchPath"].value, os.pathsep )
```
As above, I think this should be unnecessary now.

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/UniverseBlock.cpp:
```diff
@@ -62,7 +62,7 @@ void loadMetadata( const std::string &pluginPaths )
 				if( dIt->path().extension() == ".mtd" )
 				{
-					if( !AiMetaDataLoadFile( dIt->path().c_str() ) )
+					if( !AiMetaDataLoadFile( (const char*)dIt->path().c_str() ) )
```
I'm a bit confused by this one, because I expect `c_str()` to already be a `const char *`. Is it some other type on Windows, perhaps a wide char? If so, is this cast really legit?

boberfly replied:
You know what, this might definitely be the case as Arnold isn't loading in shaders from the menu currently under Windows, this might explain things...

ericmehl replied:
Yes, this is a case of Boost on Windows returning a wide string. The most straightforward solution is to just replace .c_str() with .string().c_str(). Boost does a conversion to std::string within the string() function that plain c_str() doesn't do.

I found the Arnold SDK to build against so I'm trying this out now.

johnhaddon commented on include/IECore/CamelCase.inl:
```diff
@@ -41,6 +41,11 @@
 #include "boost/algorithm/string/classification.hpp"
 #include "boost/algorithm/string/split.hpp"
 
+// MSVC gets std::toupper from <cctype>
```
This isn't unique to MSVC - we've just been lucky on Linux because `cctype` has been included coincidentally by other headers. Let's drop the `#ifdef`.

ericmehl replied:
Done and done.

ericmehl commented:
I'm going to close and then reopen this PR to trigger a new Travis build, it looks like it's passing some but not all configurations and the same code in my Travis setup is passing all configurations.

I think the only outstanding item now is to verify that Arnold is working, but that will be pending feedback from their developers regarding the [binding issue](https://github.com/ImageEngine/cortex/pull/916#discussion_r252827566) I describe above.

ericmehl commented:
Hmmmm, that didn't work as hoped. Is there another way to trigger a rebuild to see if it passes the tests?

andrewkaufman commented:
The Travis integration on Cortex is broken at the moment. We're in the process of migrating from the deprecated system to a new one.

>  Is there another way to trigger a rebuild to see if it passes the tests?

If it was working and one of the Travis iterations had failed, but you wanted to re-run it anyways (say because the failure was a bad boot), then you could do that on the associated travis page.

johnhaddon commented:
Travis should be working once again. I think you'll need to force-push or reopen the PR for it to pick up though...

ericmehl commented:
Wonderful, just got the new Travis system fired up and passing tests again.

johnhaddon commented:
Good stuff! I think that resolves everything we discussed above then?

Does anyone with expertise in the area want to comment on the CMake changes or the windows-specific code in HeaderGenerator and StreamIndexedIO? I'm not remotely qualified to review those, so in the absence of further comments I'll merge...

boberfly commented on contrib/IECoreAppleseed/src/IECoreAppleseed/outputDriver/DisplayTileCallback.cpp:
```diff
@@ -392,7 +392,8 @@ extern "C"
 // Display plugin entry point.
-IECOREAPPLESEED_API asr::ITileCallbackFactory* create_tile_callback_factory( const asr::ParamArray *params )
+IECOREAPPLESEED_API asr::ITileCallbackFactory* create_tile_callback_factory( const asr::ParamArray *params );
+asr::ITileCallbackFactory* create_tile_callback_factory( const asr::ParamArray *params )
```
Does this even compile? Seems like there shouldn't be a double-up of functions and no export here as there's already an extern "C" which should have a clean exported unmangled function for Appleseed to load in.

ericmehl replied:
Yeah, it's a little confusing with the declaration and definition right there, but it's compiling here so I'm inclined to resolve it as-is.

boberfly commented on contrib/cmake/CMakeLists.txt:
```diff
@@ -89,7 +89,7 @@ if( ${WINDOWS} )
 endif()
 
-set( CORTEX_SOURCE_DIR ${CMAKE_SOURCE_DIR}/../.. )
+set( CORTEX_SOURCE_DIR ${CMAKE_SOURCE_DIR}/../..)
```
Being pedantic here but that space on the end looks better :) can ignore

ericmehl replied:
I appreciate the consistency - fixed that one and a few others as well.

ericmehl commented:
I think that takes care of it! But definitely open to feedback on the Windows specific bits John mentioned.

ericmehl commented:
I included JOM because it builds a lot faster vs. the normal "Visual Studio 15 2017 Win64" builder (which only uses one processor). It gets the build under the 60 minute time limit of AppVeyor. An alternative if it's preferred to leave it out would be to ask the AppVeyor people for a extended limit on this project. I successfully did that for the main Gaffer build.

I lean towards leaving it in so other forks don't run into the time limit by default but I don't feel strongly about it.

boberfly commented:
Oh don't get me wrong I prefer JOM also, because I use it to build multithreaded on my docker+wine+msvc stack also. Sounds good.

ericmehl commented:
After sorting out #919 I'm inclined to push up getting all the tests running and passing to lend confidence in this PR. I'll dig into that more this weekend to get a sense of how much it's going to take to get all the tests up to speed.

ericmehl commented:
I have tests running on Windows and it looks like there are only about 100 of the 1300+ test that are failing or erroring on Windows. Unless there are objections I'm going to fix those up over the next couple weeks and add them to this PR when I'm done to have everything tied up nicely.

At a quick glance it's mostly replacing paths with forward slash with os.path.join functions.

I'll push my test changes for each module as I go in case anyone wants to review prior to having them all up and I can change course as I go along.

johnhaddon commented:
Hey @ericmehl, I'm conscious that this PR is the culmination of a lengthy effort from you and @boberfly, and as such I'm thinking we should perhaps merge it _before_ getting all tests passing on Windows. That potentially makes it easier for others to jump on board and help out with continued testing and development, rather than stalling the process waiting for perfection. What do you think?

ericmehl commented:
I'm definitely comfortable with that if you are. I have a few fixes that have come up when I was tidying up the tests that fix some broken details. I'll rebase those today or tomorrow. I can either rebase those into this PR or leave them for a follow up, whichever you prefer.

There are only a few tests left that are not passing - I'm not sure if they are revealing something benign about Windows vs. Linux differences or something that needs to be investigated more carefully. I'll open up a thread for those once I get this rebased and they will be easier for others to have a look at.

johnhaddon commented:
> I'm definitely comfortable with that if you are. I have a few fixes that have come up when I was tidying up the tests that fix some broken details. I'll rebase those today or tomorrow. I can either rebase those into this PR or leave them for a follow up, whichever you prefer.

Let's leave them as a followup - it'll be easier to reason about them separate from this. I was about to hit the big green merge button, but then in a final check I realised that "Windows Build : add AppVeyor configuration" is adding Jom binaries into the repo - I don't think I can merge that. Could you drop that commit and then we can discuss it separately?

ericmehl commented:
Great! I just rebased on the latest master branch and took out the JOM commit. It won't build on Appveyor at the moment (because of removing the commit) but I'm getting successful builds on my internal CI system.

A few things I know aren't working that I have fixed for the next PR:
- IECoreUSD is not being built
- PointDistribution (and therefore Gaffer's Seeds node) doesn't read the tileset correctly
- IFFFile doesn't read correctly
- PDCParticleReader / Writer don't read / write correctly

johnhaddon commented:
Thanks! Would be great to get the Appveyor stuff set up, so we can make sure we don't inadvertently break the Windows build in future. I just don't want to host binaries in the repo. Can you cope without Jom, or install it on Appveyor some other way? If there's anything I can do to help with the remaining test failures, just let me know...

ericmehl commented:
I'll ask the Appveyor folks about giving it an extended build time. For some reason cmake + MSVC builds single threaded but cmake + JOM builds multi-threaded and that gets it under the 60 minute time limit.

I'll likely have a few questions soon to get the remaining tests passing, there are some quirky things happening with releasing file handles in a couple tests. But the good news is it's down to only about 5 tests not passing, and even some of those are interrelated.

johnhaddon commented:
Getting the build as fast as possible is well worth it. Perhaps we can still use Jom, but download/install it from somewhere other than this repo? Probably the biggest help in keeping our Travis build times down is the build cache. This keeps object files from previous builds so that they don't need to be rebuilt if they're unaffected by the changes in the next PR. Does Appveyor have anything like that?

ericmehl commented:
I'll look into downloading Jom as part of the build. Appveyor does have a cache system but I found it not useful for speeding up the build. When it checks out the code, it does it on a fresh VM and all the timestamps for the source files are set to the time it got checked out, not the time it was committed. So even with cached build progress files, it thinks all the code is newer than what was cached and a full build is done.

Does Travis set timestamps to the commit time rather than checkout time? Or maybe they don't start from a fresh VM? I'll see if I can find anything to get around this since it's annoying to have a 40+ minute build each time.

johnhaddon commented:
> I'll look into downloading Jom as part of the build.

I've seen references to NuGet in various places - some sort of package manager. Is that something we can use?

> Does Travis set timestamps to the commit time rather than checkout time?

I don't know actually, but I don't think SCons uses timestamps - it looks at file contents instead. SCons' build cache is one of the best things about it. Is SCons a no-go on Windows? Maybe ccache is an option?

ericmehl commented:
It looks like there is a NuGet package for Jom, I'll give that a try. @boberfly - do you know about the Scons on Windows question?

boberfly commented:
Great news this got merged! :)

@ericmehl & @johnhaddon I think CMake just made things much easier with finding all the dependencies via FindXXX.cmake files than rewriting the SConstruct file for all the custom pathing that Windows needs. That and in-source builds are terrible, no offense... :)

Alternatively I'd like to experiment with seeing if the ninja build system is a possibility on windows, which should be a speed demon for building, that and trying out clang-cl which is what google uses now for chrome builds on windows, it is ABI-compatible with MSVC and links against the windows SDK directly (mingw64 isn't compatible like this so clang-cl is a real drop-in replacement) and it can run on native Linux without wine as well.

ericmehl commented:
I think I'll take a crack at adapting Scons, my head is very much is slash-world these days. It would be very nice to have quicker builds on Appveyor. I had a look at Ninja and it doesn't look like it does the content-based change tracking that Scons does. 

I'm interested to hear how you get along with clang-cl, @boberfly. I'm a wary of moving away from the most native compiler for the platform, but if it's successful for Chrome that's a solid endorsement.
